pam/go-exec/module.c

```diff
@@ -0,0 +1,929 @@
+    g_signal_connect (module_data->connection, "closed",
+                      G_CALLBACK (on_connection_closed), module_data);
+
+  return TRUE;
+}
+
+static gboolean
+setup_dbus_server (ModuleData *module_data,
+                   const char *action,
+                   GError    **error)
+{
+  g_autoptr(GDBusServer) server = NULL;
+  g_autofree char *escaped = NULL;
+  g_autofree char *server_addr = NULL;
+  g_autofree char *guid = NULL;
+  g_autofree char *tmpdir = NULL;
+
+  if (module_data->server)
+    return TRUE;
+
+  tmpdir = g_dir_make_tmp ("authd-pam-server-XXXXXX", error);
+  if (tmpdir == NULL)
+    return FALSE;
+
+  g_assert (module_data->server_tmpdir == NULL);
+  module_data->server_tmpdir = g_steal_pointer (&tmpdir);
+
+  escaped = g_dbus_address_escape_value (module_data->server_tmpdir);
+  server_addr = g_strdup_printf ("unix:tmpdir=%s", escaped);
+  guid = g_dbus_generate_guid ();
+
+  g_debug ("Setting up connection at %s (%s)", server_addr, guid);
+  server = g_dbus_server_new_sync (server_addr,
+                                   G_DBUS_SERVER_FLAGS_AUTHENTICATION_REQUIRE_SAME_USER,
+                                   guid,
+                                   NULL,
+                                   module_data->cancellable,
+                                   error);
+  if (server == NULL)
+    return FALSE;
+
+  g_dbus_server_start (server);
+  module_data->server = g_steal_pointer (&server);
+
+  g_debug ("Server started, connectable address %s",
+           g_dbus_server_get_client_address (module_data->server));
+
+  return TRUE;
+}
+
+static int
+dup_fd_checked (int fd, GError **error)
+{
+  int new_fd = dup (fd);
+
+  if (new_fd < 0)
+    {
+      int errsv = errno;
+      g_set_error_literal (error, G_IO_ERROR, g_io_error_from_errno (errsv),
+                           g_strerror (errsv));
+    }
+
+  return new_fd;
+}
+
+static gboolean
+handle_module_options (int argc, const
+                       char **argv,
+                       GPtrArray **out_args,
+                       char ***out_env_variables,
+                       GError **error)
+{
+  g_autoptr(GOptionContext) options_context = NULL;
+  g_autoptr(GPtrArray) args = NULL;
+  g_auto(GStrv) args_strv = NULL;
+  g_auto(GStrv) env_variables = NULL;
+  gboolean debug_enabled = FALSE;
+
+  const GOptionEntry options_entries[] = {
+    { "env", 'e', 0, G_OPTION_ARG_STRING_ARRAY, &env_variables, NULL, NULL },
+    { "debug", 'd', 0, G_OPTION_ARG_NONE, &debug_enabled, NULL, NULL },
+    G_OPTION_ENTRY_NULL
+  };
+
+  args = g_ptr_array_new_full (argc + 1, g_free);
+  /* We temporary add a fake item as first one, since the option parser ignores
+   * it, since normally it's just the program name */
+  g_ptr_array_add (args, g_strdup ("pam-go-exec-module"));
+  for (int i = 0; i < argc; ++i)
+    g_ptr_array_add (args, g_strdup (argv[i]));
+
+  /* FIXME: use g_ptr_array_new_null_terminated instead. */
+  g_ptr_array_add (args, NULL);
```

Comment:
Any reason why we can't already use it here?

Reply:
It's not defined in GLib for 22.04, so it would make the CI cry for not much gain... We can wait a bit more.

Reply:
I think it's worth mentioning this in the comment.

Reply:
This is done too now, and I've switch to using a `GStrvBuilder` when it's convenient.